### ISSUE TITLE
parse_example is awfully slow

### ISSUE THREAD

**Issue by balaji-in-git** [tensorflow/tensorflow#13187]
@skearnes
you have indicated in this post of yours https://github.com/tensorflow/tensorflow/issues/390 way back in 2015 that parse_example is about 30 times faster than parse_single_example.
I have tried different options to modify my simple training script which only prints about 100000 tfrecords batched in 1000 and just does a print of feature and label after session.run(feature, label). Feature is a sparseTensor BTW.
Can you please put a test sample which proved that parse_example was that fast. parse_single_example was taking ~320 secs, now parse_example takes ~240 secs.

@Admin, please do not close this issue and refer to stackoverflow, as I don't think this is something to do with API usage or wrong parameters.
This is to do with the performance of queues (enqueue, dequeue) & threads

**Comment by yaroslavvb** [comment]
It's possible there was performance regression in parse_example 

For this to get resource there needs to be at minimum a reproducible example, and ideally, some investigation of the cause of the slowness

1. It takes 240 secs, how much do you expect it to take? Is it possible that it's as as fast as it can be? Similar discussion with parse_csv slow
https://github.com/tensorflow/tensorflow/issues/11713

2.  [gperftools](https://github.com/gperftools/gperftools) profile to show where time is spent is useful, like here
https://github.com/tensorflow/tensorflow/issues/3009#issuecomment-228867287

**Comment by balaji-in-git** [comment]
just to add more information to debug, here is my lscpu

Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
Core(s) per socket:    2
Socket(s):             2
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2683 v4 @ 2.10GHz
Stepping:              1
CPU MHz:               2095.148
BogoMIPS:              4190.29
Hypervisor vendor:     VMware
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              40960K
NUMA node0 CPU(s):     0-3

**Comment by balaji-in-git** [comment]
Each TFRecord is a sparse structure of feature shape  1000 222215 viz. basically I am making batches of 1000 features. TFRecordWriter works faster as compared to burning a 1000 examples of (1x222215)  . My labels are on one-hot with 58565 labels and 1000 rows again to match 1000 labels.
Using SparseFeature and parse_example is faster compared to using parse_single_example and/or VarLenFeature. But not significantly fast. (I didn't do decode_csv, it would have been worse anyway.)
With this structure, I am able to train 400000 examples trained in batches of 1000 (each TFRecord) and 5 epochs in 861-900 secs. Is this the best I can get. I am ONLY printing the tensors. No matmuls, no optimizers.

**Comment by balaji-in-git** [comment]
That was the story with 1.2.1. I thought I should upgrade to 1.3 and now I did a bazel build hoping for better. Unfortunately, it's got worse. Same run above, now takes 1184 secs.
Does Tensorflow require some minimum hardware to show it's muscles.

**Comment by balaji-in-git** [comment]
I think the input pipeline is getting slower. I am sure if I do away with TFRecords and queues and feed the sparse matrix directly, processing using scipy or some other python library, it's been running faster earlier, though it was single threaded and may not be using all the logical cores though.

**Comment by yaroslavvb** [comment]
There are many ways to get slow performance that is not a bug in TensorFlow (ie, wrong choice of threads, wrong data format, wrong design of pipeline etc). Stackoverflow can help with that with tag "tensorflow". It could be a bug in `parse_example` op. If you think it's a bug in `parse_example`, it would help with extra details I mentioned above (reproducible example + performance profile)

**Comment by balaji-in-git** [comment]
I reduced the number of labels in the training examples from about 70000, which is the super set to about 1500 which is the number labels that's in the training set. Now the dry run, (WITHOUT matmuls, cost optimizers etc.) runs very fast. About 0.5 mn training examples pipelined and get enqueued/dequeued in about 17 secs and 5 epochs. That's about 1 million training examples in about 8-9 secs. I am not sure how & why this correlates internally, the features are SparseFeature and label is FixedLenFeature.
Now the big question, if I have to train with all the 70000 labels, will the performance again take a beating.

**Comment by balaji-in-git** [comment]
Update: I think the slowness has nothing to do with the number of labels. I removed one-hot dense vector and made it one-hot sparse and it's as fast as otherwise viz., its just a fixedLenFeature with the batch of numbers.
What helped me is to write a batch of say 1000 rows into tfrecords and read it as a single_example (though I know its a sparse-batch example) and treat it accordingly. Writing a batch of examples into tfrecords is definitely faster than writing each feature row/matrix, vector.
But parse_example is nowhere faster than parse_single_example as claimed here #390.

**Comment by hkxIron** [comment]
@balaji-in-git ,I think parse_example is much faster than parse_single_example. in my case , speed change from 25000 sample/second to 32000 sample/second

**Comment by angersson** [comment]
@balaji-in-git, can you provide code to verify this regression? Something that used to be fast but is now slow?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue? Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by michaelisard** [comment]
Closing due to lack of activity but please reopen if there's new information.
